File: modules/shv-publication-engine/tests/publication_test.cpp
#include "native_test.hpp"
using namespace native_test;
void test_native_publication_contract(Engine &e) {
  auto call = [&](std::string op, Json p, bool fail = false) {
    return e.call(op, p, !fail);
  };
  std::string h = "sha256:" + std::string(64, 'a');
  Json inst = {
      {"Role", "shv-partition-engine"},
      {"ModuleID", "shv-partition-engine"},
      {"EngineID", "symphony-shv-partition"},
      {"Version", "0.2.0-dev"},
      {"Prefix", "/partition"},
      {"ReceiptPath", "/partition/share/symphony/receipts/shv-partition-engine/"
                      "0.2.0-dev/install-receipt.json"},
      {"ReceiptDigest", h},
      {"ReceiptProtocol", "symphony.knowledge.install-receipt.v2"},
      {"ExecutablePath", "/partition/libexec/symphony/shv-partition-engine/"
                         "0.2.0-dev/symphony-shv-partition"},
      {"ExecutableDigest", h}};
  auto manifest = seal({{"protocol", "symphony.shv.partition-manifest.v1"},
                        {"entries", Json::array()},
                        {"required_references", Json::array()},
                        {"loaded_count", 0},
                        {"missing_count", 0},
                        {"complete_inventory", true},
                        {"reference_statuses", Json::array()}});
  Json d = {
      {"catalogue_id", "research"},
      {"tops_id", "00000000-0000-4000-8000-000000000001"},
      {"manifest", manifest},
      {"policy",
       {{"missing_partitions", "reject"}, {"missing_references", "reject"}}},
      {"partition_installation", inst},
      {"members", Json::array()}};
  Json base = {{"operation_id", "first"},
               {"current", nullptr},
               {"desired", d},
               {"reason", "caller selection"}};
  auto descriptor = call("inspect", Json::object());
  NT_REQUIRE(descriptor["operations"].size() == 4);
  auto plan = call("publication_plan", base);
  auto head = call("publication_reduce",
                   {{"current", nullptr}, {"plan", plan}})["head"];
  NT_REQUIRE(head["generation"] == 1);
  NT_REQUIRE(call("publication_status", {{"history", {head}}})["head"] == head);
  call("publication_plan", updated(base, {{"current", head}}), true);
  for (auto change :
       Json::array({{{"catalogue_id", "other"}},
                    {{"tops_id", "00000000-0000-4000-8000-000000000002"}}}))
    call("publication_plan",
         updated(base, {{"current", head}, {"desired", updated(d, change)}}),
         true);
  for (auto change : Json::array(
           {{{"generation", 2}}, {{"previous_digest", h}}, {{"digest", h}}})) {
    auto bad = plan;
    bad["head"].update(change);
    if (!change.contains("digest"))
      bad["head"] = seal(bad["head"]);
    call("publication_reduce", {{"current", nullptr}, {"plan", seal(bad)}},
         true);
  }
  for (auto change : Json::array({{{"members", nullptr}},
                                  {{"members", Json::array({Json::object()})}},
                                  {{"policy", Json::object()}},
                                  {{"extra", true}}}))
    call("publication_plan", updated(base, {{"desired", updated(d, change)}}),
         true);
  for (auto change : Json::array({{{"Version", "0.3.0-dev"}},
                                  {{"Prefix", "/partition/../bad"}},
                                  {{"ReceiptDigest", "x"}}}))
    call("publication_plan",
         updated(base, {{"desired", updated(d, {{"partition_installation",
                                                 updated(inst, change)}})}}),
         true);
  auto missing = seal(
      updated(manifest, {{"entries", Json::array({{{"partition_digest", h},
                                                   {"partition", nullptr}}})},
                         {"missing_count", 1},
                         {"complete_inventory", false}}));
  call("publication_plan",
       updated(base, {{"desired", updated(d, {{"manifest", missing}})}}), true);
  auto allowed = updated(d, {{"manifest", missing},
                             {"policy",
                              {{"missing_partitions", "allow"},
                               {"missing_references", "allow"}}}});
  NT_REQUIRE(call("publication_plan",
                  updated(base, {{"desired",
                                  allowed}}))["head"]["definition"] == allowed);
  Json refs = Json::array({{{"partition_digest", h}, {"subject_id", "cpu"}}});
  missing = seal(
      updated(missing, {{"required_references", refs},
                        {"reference_statuses",
                         Json::array({{{"reference", refs[0]},
                                       {"status", "missing_partition"}}})}}));
  call("publication_plan",
       updated(base,
               {{"desired",
                 updated(allowed, {{"manifest", missing},
                                   {"policy",
                                    {{"missing_partitions", "allow"},
                                     {"missing_references", "reject"}}}})}}),
       true);
  call("publication_plan",
       updated(base, {{"desired", updated(allowed, {{"manifest", missing}})}}));
  call("publication_status", {{"history", Json::array()}}, true);
  call("publication_status", {{"history", {head, head}}}, true);
  Json history = {head};
  for (int n = 2; n <= 32; ++n) {
    auto next = call(
        "publication_plan",
        {{"operation_id", "r" + std::to_string(n)},
         {"current", history.back()},
         {"desired",
          updated(d, {{"policy",
                       {{"missing_partitions", n % 2 == 0 ? "allow" : "reject"},
                        {"missing_references", "reject"}}}})},
         {"reason", "caller policy change"}});
    call("publication_reduce", {{"current", nullptr}, {"plan", next}}, true);
    history.push_back(call("publication_reduce", {{"current", history.back()},
                                                  {"plan", next}})["head"]);
  }
  NT_REQUIRE(
      call("publication_status", {{"history", history}})["history_digests"]
          .size() == 32);
  call("publication_plan", updated(base, {{"current", history.back()}}), true);
  auto overflow = history;
  overflow.push_back(history.back());
  call("publication_status", {{"history", overflow}}, true);
  for (std::string version :
       {"0.2.0-dev", "0.3.0-dev", "0.4.0-dev", "0.5.0-dev", "latest"}) {
    auto selected = inst;
    for (std::string key : {"ReceiptPath", "ExecutablePath"})
      selected[key] =
          replace(selected[key].get<std::string>(), "0.2.0-dev", version);
    selected["Version"] = version;
    bool accepted = version == "0.2.0-dev" ||
                    (descriptor["engine_version"] == "0.4.0-dev" &&
                     (version == "0.3.0-dev" || version == "0.4.0-dev"));
    call("publication_plan",
         updated(base, {{"desired",
                         updated(d, {{"partition_installation", selected}})}}),
         !accepted);
  }
}
int main(int argc, char **argv) {
  return test_main([&] {
    Arguments a(argc, argv);
    Engine e{a.require("engine"), "symphony-shv-publication",
             "publication-test"};
    test_native_publication_contract(e);
    e.summary();
  });
}

File: modules/shv-publication-engine/tests/publication_test.py
"""Pure owner conformance; synthetic references do not establish live authorization."""
import argparse, copy, hashlib, json, subprocess, time
from pathlib import Path
ENGINE=None
calls=rejects=0

def seal(v):
 v=copy.deepcopy(v);v.pop('digest',None);v['digest']='sha256:'+hashlib.sha256(json.dumps(v,sort_keys=True,separators=(',',':')).encode()).hexdigest();return v

def call(op,p,fail=False):
 global calls,rejects
 request=dict(protocol='symphony.knowledge.engine-process.v1',request_id='publication-test',correlation_id='publication-test',target_engine='symphony-shv-publication',operation=op,deadline_unix_ms=int(time.time()*1000)+30000,payload=p)
 r=subprocess.run([ENGINE],input=json.dumps(request),capture_output=True,text=True);calls+=1
 out=json.loads(r.stdout);assert bool(r.returncode)==fail,(op,r.stdout,r.stderr);assert out['outcome']==('error' if fail else 'ok');assert not r.stderr
 if fail:rejects+=1;return
 result=out['result']
 if op!='inspect':assert seal(result)==result
 return result

def test_native_publication_contract(engine):
 global ENGINE, calls, rejects
 ENGINE=engine
 calls=rejects=0
 h='sha256:'+'a'*64
 inst=dict(Role='shv-partition-engine',ModuleID='shv-partition-engine',EngineID='symphony-shv-partition',Version='0.2.0-dev',Prefix='/partition',ReceiptPath='/partition/share/symphony/receipts/shv-partition-engine/0.2.0-dev/install-receipt.json',ReceiptDigest=h,ReceiptProtocol='symphony.knowledge.install-receipt.v2',ExecutablePath='/partition/libexec/symphony/shv-partition-engine/0.2.0-dev/symphony-shv-partition',ExecutableDigest=h)
 manifest=seal(dict(protocol='symphony.shv.partition-manifest.v1',entries=[],required_references=[],loaded_count=0,missing_count=0,complete_inventory=True,reference_statuses=[]))
 d=dict(catalogue_id='research',tops_id='00000000-0000-4000-8000-000000000001',manifest=manifest,policy=dict(missing_partitions='reject',missing_references='reject'),partition_installation=inst,members=[])
 base=dict(operation_id='first',current=None,desired=d,reason='caller selection')
 descriptor=call('inspect',{});assert len(descriptor['operations'])==4
 plan=call('publication_plan',base);tr=call('publication_reduce',dict(current=None,plan=plan));head=tr['head'];assert head['generation']==1
 assert call('publication_status',dict(history=[head]))['head']==head
 call('publication_plan',{**base,'current':head},True)
 for k,v in [('catalogue_id','other'),('tops_id','00000000-0000-4000-8000-000000000002')]:call('publication_plan',{**base,'current':head,'desired':{**d,k:v}},True)
 for k,v in [('generation',2),('previous_digest',h),('digest',h)]:
  bad=copy.deepcopy(plan);bad['head'][k]=v
  if k!='digest':bad['head']=seal(bad['head'])
  call('publication_reduce',dict(current=None,plan=seal(bad)),True)
 for k,v in [('members',None),('members',[{}]),('policy',{}),('extra',True)]:call('publication_plan',{**base,'desired':{**d,k:v}},True)
 for k,v in [('Version','0.3.0-dev'),('Prefix','/partition/../bad'),('ReceiptDigest','x')]:call('publication_plan',{**base,'desired':{**d,'partition_installation':{**inst,k:v}}},True)
 missing=seal({**manifest,'entries':[dict(partition_digest=h,partition=None)],'missing_count':1,'complete_inventory':False})
 call('publication_plan',{**base,'desired':{**d,'manifest':missing}},True)
 allowed={**d,'manifest':missing,'policy':dict(missing_partitions='allow',missing_references='allow')}
 assert call('publication_plan',{**base,'desired':allowed})['head']['definition']==allowed
 refs=[dict(partition_digest=h,subject_id='cpu')]
 missing=seal({**missing,'required_references':refs,'reference_statuses':[{'reference':refs[0],'status':'missing_partition'}]})
 call('publication_plan',{**base,'desired':{**allowed,'manifest':missing,'policy':dict(missing_partitions='allow',missing_references='reject')}},True)
 call('publication_plan',{**base,'desired':{**allowed,'manifest':missing}})
 call('publication_status',dict(history=[]),True)
 call('publication_status',dict(history=[head,head]),True)
 history=[head]
 for n in range(2,33):
  desired={**d,'policy':dict(missing_partitions='allow' if n%2==0 else 'reject',missing_references='reject')}
  nxt=call('publication_plan',dict(operation_id='r'+str(n),current=history[-1],desired=desired,reason='caller policy change'))
  call('publication_reduce',dict(current=None,plan=nxt),True)
  history.append(call('publication_reduce',dict(current=history[-1],plan=nxt))['head'])
 assert len(call('publication_status',dict(history=history))['history_digests'])==32
 call('publication_plan',{**base,'current':history[-1]},True)
 call('publication_status',dict(history=history+[history[-1]]),True)
 for version in ['0.2.0-dev','0.3.0-dev','0.4.0-dev','0.5.0-dev','latest']:
  selected=copy.deepcopy(inst)
  for key in ['ReceiptPath','ExecutablePath']:selected[key]=selected[key].replace('0.2.0-dev',version)
  selected['Version']=version
  call('publication_plan',{**base,'desired':{**d,'partition_installation':selected}},not (version=='0.2.0-dev' or (descriptor['engine_version']=='0.4.0-dev' and version in ['0.3.0-dev','0.4.0-dev'])))
 print(json.dumps(dict(status='passed',calls=calls,rejections=rejects)))


if __name__ == "__main__":
 p=argparse.ArgumentParser();p.add_argument("--engine",required=True);a=p.parse_args()
 test_native_publication_contract(a.engine)
